### PR TITLE
Improve performance of  "in" condition on some version of MySQl

### LINKED ISSUE #25199: Catalog Category Indexing takes very long on MariaDB 10.3 with many products
### Description (*)

I know MariaDB 10.3 is not yet supported, as per dev docs, so I add this as feature requests, while it will manifest as a bug as soon as we support MariaDB 10.3

Magento should remove the quotes in the SELECT entity_id IN statement when creating indexes. The entity_ids are not strings, and this seems to confuse the optimizer.

### Steps to reproduce:

1. Change value for simple products to 50000 for example (setup/performance-toolkit/profiles/ce/small.xml);
2. Run in console: bin/magento setup:perf:generate-fixtures /var/www/html/magento24/setup/performance-toolkit/profiles/ce/small.xml for example

### Expected behavior (*)

Catalog Category Indexing should run quickly, on MariaDB 10.1 and on 10.3.18

### Benefits
Shop ownes can use MariaDB 10.3.18

### Additional information

See this answer on StackExchange: https://magento.stackexchange.com/a/293345/81

I took the full query which is run several times (for each store view) on the catalog category products indexer and extracted the select part

	SELECT 2 AS `category_id`, `cp`.`entity_id` AS `product_id`, IF(ccp.product_id IS NOT NULL, ccp.position, 0) AS `position`, IF(ccp.product_id IS NOT NULL, 1, 0) AS `is_parent`, 2 AS `store_id`, IFNULL(cpvs.value, cpvd.value) AS `visibility` FROM `catalog_product_entity` AS `cp`
	 INNER JOIN `catalog_product_website` AS `cpw` ON cpw.product_id = cp.entity_id
	 INNER JOIN `catalog_product_entity_int` AS `cpsd` ON cpsd.entity_id = cp.entity_id AND cpsd.store_id = 0 AND cpsd.attribute_id = 97
	 LEFT JOIN `catalog_product_entity_int` AS `cpss` ON cpss.entity_id = cp.entity_id AND cpss.attribute_id = cpsd.attribute_id AND cpss.store_id = 2
	 INNER JOIN `catalog_product_entity_int` AS `cpvd` ON cpvd.entity_id = cp.entity_id AND cpvd.store_id = 0 AND cpvd.attribute_id = 99
	 LEFT JOIN `catalog_product_entity_int` AS `cpvs` ON cpvs.entity_id = cp.entity_id AND cpvs.attribute_id = cpvd.attribute_id  AND cpvs.store_id = 2
	 LEFT JOIN `catalog_category_product` AS `ccp` ON ccp.product_id = cp.entity_id WHERE (cpw.website_id = '1') AND (IFNULL(cpss.value, cpsd.value) = 1) AND           (IFNULL(cpvs.value, cpvd.value) IN (2, 3, 4)) AND (cp.entity_id IN ('107280', '107281', '107282', '107283', '107284'
	    [... around 70 K entity IDs ...]
	    187965', '187966')) GROUP BY `cp`.`entity_id`;

This query alone runs 8 minutes on Maria DB 10.3

When I simply remove all the quotes via s/'//g in vi in the line with the "IN" statement I get

	SELECT 2 AS `category_id`, `cp`.`entity_id` AS `product_id`, IF(ccp.product_id IS NOT NULL, ccp.position, 0) AS `position`, IF(ccp.product_id IS NOT NULL, 1, 0) AS `is_parent`, 2 AS `store_id`, IFNULL(cpvs.value, cpvd.value) AS `visibility` FROM `catalog_product_entity` AS `cp`
	 INNER JOIN `catalog_product_website` AS `cpw` ON cpw.product_id = cp.entity_id
	 INNER JOIN `catalog_product_entity_int` AS `cpsd` ON cpsd.entity_id = cp.entity_id AND cpsd.store_id = 0 AND cpsd.attribute_id = 97
	 LEFT JOIN `catalog_product_entity_int` AS `cpss` ON cpss.entity_id = cp.entity_id AND cpss.attribute_id = cpsd.attribute_id AND cpss.store_id = 2
	 INNER JOIN `catalog_product_entity_int` AS `cpvd` ON cpvd.entity_id = cp.entity_id AND cpvd.store_id = 0 AND cpvd.attribute_id = 99
	 LEFT JOIN `catalog_product_entity_int` AS `cpvs` ON cpvs.entity_id = cp.entity_id AND cpvs.attribute_id = cpvd.attribute_id  AND cpvs.store_id = 2
	 LEFT JOIN `catalog_category_product` AS `ccp` ON ccp.product_id = cp.entity_id WHERE (cpw.website_id = '1') AND (IFNULL(cpss.value, cpsd.value) = 1) AND           (IFNULL(cpvs.value, cpvd.value) IN (2, 3, 4)) AND (cp.entity_id IN (107280, 107281, 107282, 107283, 107284
	    [... around 70 K entity IDs ...]
	    187965, 187966)) GROUP BY `cp`.`entity_id`;

### Next steps

I will try to provide a patch.

### PR DESCRIPTION
**Pull request by kandy**
Improve performance of  "in" condition on some version of MySql

### Related Pull Requests

Workaround https://jira.mariadb.org/browse/MDEV-20900

### Fixed Issues (if relevant)

1. Fixes magento/magento2#25199 :Catalog Category Indexing takes very long on MariaDB 10.3 with many products

### Manual testing scenarios (*)

see #25199

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php [hunk 1]
```diff
@@ -84,7 +84,7 @@ protected function _renderFilters()
             $parentIds[] = $product->getData($metadata->getLinkField());
         }
 
-        $this->getSelect()->where('link_table.parent_id in (?)', $parentIds);
+        $this->getSelect()->where('link_table.parent_id in (?)', $parentIds,\Zend_Db::BIGINT_TYPE);
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]:
```suggestion
        $this->getSelect()->where('link_table.parent_id in (?)', $parentIds, \Zend_Db::BIGINT_TYPE);
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kandy,
Your changes looks good, except one small missing space.
Due to definition of done all changes should be covered with some tests. Could you cover your changes with any type of test?

**Comment by ihor-sviziev** [comment]
@kandy I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by ihor-sviziev** [comment]
I just re-visited this PR, looks really important to fix it, while covering original issue this with any type of automated tests will be almost impossible.
Now waiting for test results and will approve it

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kandy,
Could you fix static tests and integration tests for ee and b2b versions?

File: app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php [hunk 2]
```diff
@@ -209,10 +217,10 @@ private function retrieveAffectedIds(array $skus)
         $affectedIds = [];
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php:217]:
```suggestion
        $affectedIds = [[]];
```
To prevent issue when no products ids will be returned.
It’s needed only because of added `array_merge(...$affectedIds)`

File: dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php [hunk 3]
```diff
@@ -53,6 +53,7 @@ public function getRealMemoryUsage()
             // fall back to the Unix command line
             $result = $this->_getUnixProcessMemoryUsage($pid);
         } catch (\Magento\Framework\Exception\LocalizedException $e) {
+            echo  $e;
```
Comment (ihor-sviziev) [inline dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php]: Looks like it should be reverted

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kandy,
I just reviewed test results - looks like your changes caused some regression issue during reindex, you can see some sql error in integration tests. Please review and adjust your solution

**Review by ihor-sviziev — DISMISSED** [review]
Approved ✔
Failing functional tests not related to changes from this PR.
Failing SVC looks false positive.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7542](https://jira.corp.magento.com/browse/ENGCOM-7542) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @kandy .

Unfortunately, we are not able to reproduce this [issue](https://github.com/magento/magento2/issues/25199) on fresh 2.4-develop

**_Preconditions:_**
1. magento 2.4-develop
2. php 7.3
3. MariaDB 10.3.23
4. Have 100 categories and 100000 products for example

**Manual testing scenario:**

Run in console: **php bin/magento indexer:reindex**

**Actual Result:** : :heavy_check_mark: The indexing time is 18 sec

screenshot_29

_Note:_ With changes in PR, the indexing result is the same as in the mainline

@kandy Maybe we missed something. Could you provide us additional information on how to reproduce the problem?

Thanks!

**Comment by ihor-sviziev** [comment]
@engcom-Alfa the issue was fixed in MariaDB 10.3.22, so you need to test on earlier versions. https://jira.mariadb.org/browse/MDEV-20871

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7542](https://jira.corp.magento.com/browse/ENGCOM-7542) has been created to process this Pull Request

File: app/code/Magento/Catalog/Model/Indexer/Category/Product/Action/Full.php [hunk 4]
```diff
@@ -174,7 +174,7 @@ protected function reindex(): void
         foreach ($this->storeManager->getStores() as $store) {
             if ($this->getPathFromCategoryId($store->getRootCategoryId())) {
                 $userFunctions[$store->getId()] = function () use ($store) {
-                    return $this->reindexStore($store);
+                    $this->reindexStore($store);
```
Comment (tmotyl) [inline app/code/Magento/Catalog/Model/Indexer/Category/Product/Action/Full.php:177]: is this change related to this pr ?

**Comment by tmotyl** [comment]
It would be much easier to review the change code style changes were excluded (like adding "use" statements).

File: app/code/Magento/Catalog/Model/Indexer/Category/Product/Action/Full.php (hunk 4 above)
Comment (kandy) [reply app/code/Magento/Catalog/Model/Indexer/Category/Product/Action/Full.php:177]: This suggestion by PHPStan. "reindexStore" return type is defined as void, so there needs to return result

**Comment by kandy** [comment]
> It would be much easier to review the change code style changes were excluded (like adding "use" statements).
It would be even easier for me, but the static test crashes because it only checks for updated files

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Alfa as we discussed in Slack - looks like this PR isn't solving the described issue, indexation still works really slow.
Could you add more info and would be nice to see which SQLs are running for a long time during reindex?

**Comment by tmotyl** [comment]
@ihor-sviziev I think you should test performance of this patch with my patch applied too  https://github.com/magento/magento2/pull/27980/
As this patch did not solve the issue with  casting when type is INT_TYPE) see line
 https://github.com/magento/magento2/pull/27980/files#diff-fe16896d208c3ab2d52fccbbf7a5eaebR119

**Comment by ihor-sviziev** [comment]
@tmotyl good to know.
@engcom-Alfa could you check changes from these two PRs together?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@engcom-Alfa sent me in Slack that he checked these 2 PRs together (with #27980 ), but unfortunately it's still not fixing the issue.

He tested on MariaDB 10.3.18 with 50000 simple products.

Here is screenshots:
image
image


@engcom-Alfa could you please copy and paste here full SQL that running too long? It will be really useful for investigation.

@kandy looks like your changes not fully fixes the issue. Could you try to reproduce it again and update your PR accordingly? 

Thank you!

**Comment by ihor-sviziev** [comment]
@kandy  Just an idea  - maybe we can write in the docs that those buggy versions of Maria DB  isn't supported rather than fixing it? What do you thing?

**Comment by tmotyl** [comment]
eh, can we at last merge the change https://github.com/magento/magento2/pull/27980/files which only fixes the bug without adding casting everywhere?
I still believe that proper casting should be added (not sure whether this should be \Zend_Db::BIGINT_TYPE) or Zend_Db::INT_TYPE is enough).

**Comment by ihor-sviziev** [comment]
@tmotyl I don't mind to merge these changes that @kandy did, it's still good improvement, but it's still not fixing the issue.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by kandy** [comment]
```
~/w/perf ❯❯❯ bin/mysql                                                                                                                                                 master ✭ ✖ ✱ ◼
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 28
Server version: 10.3.18-MariaDB-1:10.3.18+maria~bionic mariadb.org binary distribution

bash-5.0# bin/magento indexer:reset catalog_category_product &&  bin/magento indexer:reindex  catalog_category_product
Category Products indexer has been invalidated.
Category Products index has been rebuilt successfully in 00:02:24
Catalog Search index has been rebuilt successfully in 00:01:10
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7542](https://jira.corp.magento.com/browse/ENGCOM-7542) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Magento 2.4-develop
MariaDB v.10.3.18

Screenshot from 2020-08-05 13-49-00

**Manual testing scenario:**
1. Change value for simple products to 50000 for example (**setup/performance-toolkit/profiles/ce/small.xml**);
2. Run in console: `bin/magento setup:perf:generate-fixtures /var/www/html/magento24/setup/performance-toolkit/profiles/ce/small.xml
` for example

**Before:** :heavy_multiplication_x: Indexation works very slow, freezes and does not pass...

Screenshot from 2020-08-05 13-47-39

**After:** :heavy_check_mark:  Catalog Category Indexing run quickly. For 50,000 products, it took about 3 minutes.

Screenshot from 2020-08-05 13-09-21

and all products were successfully indexed

**Comment by engcom-Charlie** [comment]
It needs to approve because **Semantic Version Checker** falls with errors:
* Magento\Eav\Model\ResourceModel\Entity\Attribute\Collection::_afterLoadData **[protected] Method return typing changed**.
* Magento\Catalog\Model\ResourceModel\Product\Link\Product\Collection::getLinkModel **[public] Method return typing changed.**

**Comment by ihor-sviziev** [comment]
@slavvka @lenaorobei @VladimirZaets how we can speed up approval process for SVC failure for this PR? This is really important fix.

**Comment by sidolov** [comment]
Hi @ihor-sviziev , we will proceed with changes approval ASAP

File: app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php [hunk 5]
```diff
@@ -206,13 +214,13 @@ private function getAttributeId()
      */
     private function retrieveAffectedIds(array $skus)
     {
-        $affectedIds = [];
+        $affectedIds = [[]];
```
Comment (gabrieldagama) [inline app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php]: Do we need a multi-dimensional array here?

File: app/code/Magento/CatalogRuleConfigurable/Plugin/CatalogRule/Model/ConfigurableProductsProvider.php [hunk 6]
```diff
@@ -27,20 +27,22 @@ public function __construct(\Magento\Framework\App\ResourceConnection $resource)
...
     /**
+     * Return list of ID for product variation
+     *
      * @param array $ids
      * @return array
      */
     public function getIds(array $ids)
     {
-        $key =  md5(json_encode($ids));
+        $key = md5(json_encode($ids)); //phpcs:ignore
```
Comment (gabrieldagama) [inline app/code/Magento/CatalogRuleConfigurable/Plugin/CatalogRule/Model/ConfigurableProductsProvider.php:37]: What are we ignoring here with `//phpcs:ignore`?

File: app/code/Magento/Bundle/Model/ResourceModel/Selection/Collection.php [hunk 7]
```diff
@@ -215,7 +215,7 @@ public function joinPrices($websiteId)
     public function setOptionIdsFilter($optionIds)
     {
         if (!empty($optionIds)) {
-            $this->getSelect()->where('selection.option_id IN (?)', $optionIds);
+            $this->getSelect()->where('selection.option_id IN (?)', $optionIds, \Zend_Db::BIGINT_TYPE);
```
Comment (gabrieldagama) [inline app/code/Magento/Bundle/Model/ResourceModel/Selection/Collection.php]: Is there a reason to be using Zend_Db::BIGINT_TYPE instead of Zend_Db::INT_TYPE? I guess all fields that we are using on the where the condition is an integer.

File: app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php [hunk 8]
```diff
@@ -426,7 +439,7 @@ protected function _addSetInfo()
     /**
      * Ad information about attribute sets to collection result data
      *
-     * @return \Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection
+     * @return $this
```
Comment (gabrieldagama) [inline app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php]:
```suggestion
     * @return \Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection
```

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @kandy, thanks for your pull request, it looks great! Please follow below some small questions regards the changes, let me know if you have any question.

**Comment by gabrieldagama** [comment]
@magento run all tests

File: app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php (hunk 8 above)
Comment (tmotyl) [reply app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php]: Why?

File: app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Product/Price/PricePersistence.php]: It's needed for getting rid of array_marge in the loop, but we can actually have there just  `$affectedIds = [];`, and below something like ` $affectedIds = array_merge([], ...$affectedIds);` that will look a bit better

File: app/code/Magento/CatalogRuleConfigurable/Plugin/CatalogRule/Model/ConfigurableProductsProvider.php (hunk 6 above)
Comment (ihor-sviziev) [reply app/code/Magento/CatalogRuleConfigurable/Plugin/CatalogRule/Model/ConfigurableProductsProvider.php:37]: Static tests are failing due to using md5, that we didn't added in this PR. Changing it to some other algorithm might cause some regression. It's not purpose of this PR of fixing using md5 there.

File: app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php (hunk 8 above)
Comment (ihor-sviziev) [reply app/code/Magento/Eav/Model/ResourceModel/Entity/Attribute/Collection.php]: I believe because it's still returns value from parent class, where declared `\Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection` instead of returning `$this`. I don't mind for reverting this change

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Magento 2.4-develop
MariaDB v.10.3.18

Screenshot from 2020-08-05 13-49-00

**Manual testing scenario:**
1. Change value for simple products to 50000 for example (**setup/performance-toolkit/profiles/ce/small.xml**);
2. Run in console: `bin/magento setup:perf:generate-fixtures /var/www/html/magento24/setup/performance-toolkit/profiles/ce/small.xml
` for example

**Before:** :heavy_multiplication_x: Indexation works very slow, freezes and does not pass...

Screenshot from 2020-08-05 13-47-39

**After:** :heavy_check_mark:  Catalog Category Indexing run quickly. For 50,000 products, it took about 3 minutes.



and all products were successfully indexed in about 2 minutes

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Finally it got merged! 🎉

**Comment by Quazz** [comment]
Glad to see this get merged!

I found 2 more instances of this in the catalog_url_rewrite module. I think the reason it wasn't in this PR is because it doesn't go through the DB/Select but instead it is about creating a temporary table. 

https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogUrlRewrite/Model/Map/DataCategoryUrlRewriteDatabaseMap.php#L91

That one especially can take a long time to get it done.

It can be wrapped in an array_map('intval', array_merge(...)) in order to force the productids to come through as INT

**Comment by ihor-sviziev** [comment]
@Quazz feel free to create PR with your suggested changes
